contrib/twitchtv/twirp/twirp.go

```diff
@@ -0,0 +1,207 @@
+		opts = append(opts, tracer.ChildOf(spanctx))
+	}
+
+	span, ctx := tracer.StartSpanFromContext(req.Context(), "twirp.request", opts...)
+	defer span.Finish()
+
+	err := tracer.Inject(span.Context(), tracer.HTTPHeadersCarrier(req.Header))
+	if err != nil {
+		log.Warn("contrib/twitchtv/twirp.wrappedClient: failed to inject http headers: %v\n", err)
+	}
+
+	req = req.WithContext(ctx)
+	res, err := wc.c.Do(req)
+	if err != nil {
+		span.SetTag(ext.Error, err)
+	} else {
+		span.SetTag(ext.HTTPCode, strconv.Itoa(res.StatusCode))
+		// treat 4XX and 5XX as errors for a client
+		if res.StatusCode >= 400 && res.StatusCode < 600 {
+			span.SetTag(ext.Error, fmt.Errorf("%s", res.Status))
+		}
+	}
+	return res, err
+}
+
+// WrapServer wraps a http.Handler to add distributed tracing to a Twirp server.
+func WrapServer(h http.Handler, opts ...Option) http.Handler {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		opts := []tracer.StartSpanOption{
+			tracer.SpanType(ext.SpanTypeWeb),
+			tracer.Tag(ext.HTTPMethod, r.Method),
+			tracer.Tag(ext.HTTPURL, r.URL.Path),
+		}
+		if cfg.serviceName != "" {
+			opts = append(opts, tracer.ServiceName(cfg.serviceName))
+		}
+		if !math.IsNaN(cfg.analyticsRate) {
+			opts = append(opts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+		}
+		if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header)); err == nil {
+			opts = append(opts, tracer.ChildOf(spanctx))
+		}
+
+		span, ctx := tracer.StartSpanFromContext(r.Context(), "twirp.request", opts...)
+		defer span.Finish()
+
+		r = r.WithContext(ctx)
+		h.ServeHTTP(w, r)
+	})
+}
+
+// NewServerHooks creates the callback hooks for a twirp server to perform tracing.
+// It is used in conjunction with WrapServer.
+func NewServerHooks(opts ...Option) *twirp.ServerHooks {
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	return &twirp.ServerHooks{
+		RequestReceived:  requestReceivedHook(cfg),
+		RequestRouted:    requestRoutedHook(cfg),
+		ResponsePrepared: responsePreparedHook(cfg),
+		ResponseSent:     responseSentHook(cfg),
+		Error:            errorHook(cfg),
+	}
+}
+
+func requestReceivedHook(cfg *config) func(context.Context) (context.Context, error) {
+	return func(ctx context.Context) (context.Context, error) {
+		opts := []tracer.StartSpanOption{
+			tracer.SpanType(ext.SpanTypeWeb),
+		}
+		if cfg.serviceName != "" {
+			opts = append(opts, tracer.ServiceName(cfg.serviceName))
+		}
+		if pkg, ok := twirp.PackageName(ctx); ok {
+			opts = append(opts, tracer.Tag("twirp.package", pkg))
+		}
+		if svc, ok := twirp.ServiceName(ctx); ok {
+			opts = append(opts, tracer.Tag("twirp.service", svc))
+		}
+		if !math.IsNaN(cfg.analyticsRate) {
+			opts = append(opts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+		}
+		_, ctx = tracer.StartSpanFromContext(ctx, "twirp.request", opts...)
```

Comment:
Can we do better with the operation name here? Maybe `twirp.received` or `twirp.request.received` or `request.received`. I'm thinking to have something different. Its weird looking at it to have the same name.